### PR TITLE
[Clang][AArch64] Fix Pure Scalables Types argument passing and return

### PR DESCRIPTION
**Pull request by momchil-velikov**
Pure Scalable Types are defined in AAPCS64 here:
  https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#pure-scalable-types-psts

And should be passed according to Rule C.7 here:
  https://github.com/ARM-software/abi-aa/blob/main/aapcs64/aapcs64.rst#682parameter-passing-rules

This part of the ABI is completely unimplemented in Clang, instead it treats PSTs sometimes as HFAs/HVAs, sometime as general composite types.

This patch implements the rules for passing PSTs by employing the `CoerceAndExpand` method and extending it to:
  * allow array types in the `coerceToType`; Now only `[N x i8]` are considered padding.
  * allow mismatch between the elements of the `coerceToType` and the elements of the `unpaddedCoerceToType`; AArch64 uses this to map fixed-length vector types to SVE vector types.

Corectly passing a PST argument needs a decision in Clang about whether to pass it in memory or registers or, equivalently, whether to use the `Indirect` or `Expand/CoerceAndExpand` method. It was considered relatively harder (or not practically possible) to make that decision in the AArch64 backend.
Hence this patch implements the register counting from AAPCS64 (cf. `NSRN`, `NPRN`) to guide the Clang's decision.

### CODE REVIEW

File: clang/lib/CodeGen/Targets/AArch64.cpp [hunk 1]
```diff
@@ -423,6 +510,19 @@ ABIArgInfo AArch64ABIInfo::classifyReturnType(QualType RetTy,
     // Homogeneous Floating-point Aggregates (HFAs) are returned directly.
     return ABIArgInfo::getDirect();
 
+  // In AAPCS return values of a Pure Scalable type are treated is a first named
```
Comment (jthackray) [inline clang/lib/CodeGen/Targets/AArch64.cpp:513]: nit: s/are treated is a/are treated as a/ ?

File: clang/lib/CodeGen/Targets/AArch64.cpp [hunk 2]
```diff
@@ -533,11 +638,158 @@ bool AArch64ABIInfo::isZeroLengthBitfieldPermittedInHomogeneousAggregate()
...
+    if (CGCXXABI::RecordArgABI RAA = getRecordArgABI(RT, getCXXABI());
+        RAA != CGCXXABI::RAA_Default)
+      return false;
+
+    // Pure scalable types are never unions and never contain unions.
+    const RecordDecl *RD = RT->getDecl();
+    if (RD->isUnion())
+      return false;
+
+    // If this is a C++ record, check the bases bases.
```
Comment (jthackray) [inline clang/lib/CodeGen/Targets/AArch64.cpp:679]: Typo?

File: clang/lib/CodeGen/Targets/AArch64.cpp [hunk 3]
```diff
@@ -533,11 +638,158 @@ bool AArch64ABIInfo::isZeroLengthBitfieldPermittedInHomogeneousAggregate()
...
+  Flattened.push_back(Ty);
+}
+
 RValue AArch64ABIInfo::EmitAAPCSVAArg(Address VAListAddr, QualType Ty,
                                       CodeGenFunction &CGF, AArch64ABIKind Kind,
                                       AggValueSlot Slot) const {
-  ABIArgInfo AI = classifyArgumentType(Ty, /*IsVariadic=*/true,
-                                       CGF.CurFnInfo->getCallingConvention());
+  // These numbers are not used for variadic arguments, hence it doesn't matter
+  // they don't retain their values accross multiple calls to
```
Comment (jthackray) [inline clang/lib/CodeGen/Targets/AArch64.cpp:787]: nit: s/accross/across/

File: clang/lib/CodeGen/Targets/AArch64.cpp [hunk 4]
```diff
@@ -533,11 +638,158 @@ bool AArch64ABIInfo::isZeroLengthBitfieldPermittedInHomogeneousAggregate()
...
+
+    // Pure scalable types are never unions and never contain unions.
+    const RecordDecl *RD = RT->getDecl();
+    if (RD->isUnion())
+      return false;
+
+    // If this is a C++ record, check the bases bases.
+    if (const CXXRecordDecl *CXXRD = dyn_cast<CXXRecordDecl>(RD)) {
+      for (const auto &I : CXXRD->bases()) {
+        if (isEmptyRecord(getContext(), I.getType(), true))
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/Targets/AArch64.cpp:682]: Does this handle empty unions correctly?

File: clang/lib/CodeGen/Targets/AArch64.cpp [hunk 5]
```diff
@@ -533,11 +638,158 @@ bool AArch64ABIInfo::isZeroLengthBitfieldPermittedInHomogeneousAggregate()
...
+    return true;
+  }
+
+  const auto *VT = Ty->getAs<VectorType>();
+  if (!VT)
+    return false;
+
+  if (VT->getVectorKind() == VectorKind::SveFixedLengthPredicate) {
+    ++NPred;
+    if (CoerceToSeq.size() < 12)
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/Targets/AArch64.cpp:707]: If CoerceToSeq is too large, do you not just want to return false here?

File: clang/lib/CodeGen/Targets/AArch64.cpp [hunk 6]
```diff
@@ -353,6 +427,17 @@ AArch64ABIInfo::classifyArgumentType(QualType Ty, bool IsVariadic,
         nullptr, true, Align);
   }
 
+  // In AAPCS named arguments of a Pure Scalable Type are passed expanded in
+  // registers, or indirectly if there are not enough registers.
+  if (Kind == AArch64ABIKind::AAPCS && !IsVariadic) {
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/Targets/AArch64.cpp:432]: What spec rule excludes variadic functions?

**Review by efriedma-quic — COMMENTED** [review]
Clang tracks the number of registers used on other targets, so it's not unprecedented... just an unfortunate consequence of the limitations of ABI in LLVM IR.

File: clang/lib/CodeGen/Targets/AArch64.cpp (hunk 6 above)
Comment (momchil-velikov) [reply clang/lib/CodeGen/Targets/AArch64.cpp:432]: That's Rule C.7, that wants the argument to be named.

> If the argument is a Pure Scalable Type that consists of NV Scalable Vector Types and
> NP Scalable Predicate Types, **if the argument is named**, if NSRN+NV ≤ 8,
> and if NPRN+NP ≤ 4, then the Scalable Vector Types are

File: clang/lib/CodeGen/Targets/AArch64.cpp (hunk 4 above)
Comment (momchil-velikov) [reply clang/lib/CodeGen/Targets/AArch64.cpp:682]: Turns out there's a bug a few lines down
```
    // Check members.
    for (const auto *FD : RD->fields()) {
      QualType FT = FD->getType();
      if (isEmptyRecord(getContext(), FT, true))
        continue;
...
```

There should be a call to `isEmptyField`.

Comment (momchil-velikov) [reply clang/lib/CodeGen/Targets/AArch64.cpp:687]: Fixed, added test.

File: clang/lib/CodeGen/Targets/AArch64.cpp (hunk 5 above)
Comment (momchil-velikov) [reply clang/lib/CodeGen/Targets/AArch64.cpp]: Done.

File: clang/lib/CodeGen/Targets/AArch64.cpp (hunk 6 above)
Comment (efriedma-quic) [reply clang/lib/CodeGen/Targets/AArch64.cpp:432]: I'm pretty sure "IsVariadic" is whether the function type is variadic, not whether the argument is named.

Comment (momchil-velikov) [reply clang/lib/CodeGen/Targets/AArch64.cpp]: Indeed! Fixed.

File: clang/test/CodeGen/aarch64-pure-scalable-args.c [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+//   p->a    -> p0
+//   p->x    -> q0
+//   p->y[0] -> q1
+//   p->y[1] -> q2
+//   p->z    -> q3
+//   p->b    -> p1
+//   *q -> memory, address -> x1
+void test_pass_variadic(PST *p, PST *q) {
+    void pass_variadic_callee(PST, ...);
+    pass_variadic_callee(*p, *q);
```
Comment (efriedma-quic) [inline clang/test/CodeGen/aarch64-pure-scalable-args.c:303]: Please also CHECK the call instruction.

Comment (efriedma-quic) [reply clang/test/CodeGen/aarch64-pure-scalable-args.c:303]: Also, I'd like to see a test for va_arg.

Comment (momchil-velikov) [reply clang/test/CodeGen/aarch64-pure-scalable-args.c:303]: Will do. Variadic arg passing is somewhat broken right now, gonna fix that too.

File: clang/lib/CodeGen/Targets/AArch64.cpp [hunk 8]
```diff
@@ -303,6 +368,20 @@ AArch64ABIInfo::classifyArgumentType(QualType Ty, bool IsVariadic,
...
         return getNaturalAlignIndirect(Ty, false);
 
+    if (const BuiltinType *BT = Ty->getAs<BuiltinType>()) {
+      if (BT->isSVEBool() || BT->isSVECount())
+        NPRN = std::min(NPRN + 1, 4u);
+      else if (BT->getKind() == BuiltinType::SveBoolx2)
+        NPRN = std::min(NPRN + 2, 4u);
+      else if (BT->getKind() == BuiltinType::SveBoolx4)
+        NPRN = std::min(NPRN + 4, 4u);
+      else if (BT->isFloatingPoint() || BT->isVectorType())
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/Targets/AArch64.cpp:378]: `BT->isVectorType()` is always false, by definition: it's just `isa<VectorType>()`, and a BuiltinType can't be a VectorType.

Comment (momchil-velikov) [reply clang/lib/CodeGen/Targets/AArch64.cpp]: Fixed and tests added.

**Review by jthackray — APPROVED** [review]
LGTM

**Review by efriedma-quic — APPROVED** [review]

**Comment by momchil-velikov** [comment]
Thanks for the reviews, much appreciated!

**Pull request merged**